File: isorpg/ts/tests/harness.ts
// 세 언어 공통의 아주 작은 테스트 하네스 (파이썬 py/tests/harness.py 와 같은 출력).
//
// 파이썬 쪽은 파일 하나가 곧 프로세스 하나라 done() 에서 sys.exit 를 부른다.
// 노드에서는 한 프로세스가 열두 모듈을 이어 돌리는 편이 훨씬 빠르므로
// done() 은 실패 수만 돌려주고, 종료 코드는 run.ts 가 마지막에 한 번만 정한다.
import * as fs from 'fs';
import * as path from 'path';

// dist/tests -> dist -> ts -> isorpg
export const BASE = path.resolve(__dirname, '..', '..', '..');
export const GOLDEN = path.join(BASE, 'golden');

let okCount = 0;
let badCount = 0;
let curName = '?';

export function title(name: string): void {
  curName = name;
  okCount = 0;
  badCount = 0;
  console.log('== ' + name + ' ==');
}

/** 파이썬의 %r 자리. 배열·객체까지 눈으로 견줄 수 있으면 충분하다. */
export function show(v: unknown): string {
  if (typeof v === 'string') return JSON.stringify(v);
  if (v instanceof Uint8Array) return '[' + Array.from(v).join(',') + ']';
  if (Array.isArray(v)) return '[' + v.map(show).join(', ') + ']';
  return String(v);
}

/** 파이썬의 == 는 리스트·튜플을 값으로 견준다. JS 의 === 는 참조를 견주므로
 *  구조 비교를 직접 만든다. 이것이 이식에서 가장 자주 새는 구멍이다. */
export function deepEq(a: unknown, b: unknown): boolean {
  if (a === b) return true;
  const au = a instanceof Uint8Array;
  const bu = b instanceof Uint8Array;
  if (au || bu) {
    if (!au || !bu) return false;
    if (a.length !== b.length) return false;
    for (let i = 0; i < a.length; i++) if (a[i] !== b[i]) return false;
    return true;
  }
  if (Array.isArray(a) && Array.isArray(b)) {
    if (a.length !== b.length) return false;
    for (let i = 0; i < a.length; i++) if (!deepEq(a[i], b[i])) return false;
    return true;
  }
  return false;
}

export function check(what: string, got: unknown, want: unknown): boolean {
  if (deepEq(got, want)) {
    okCount += 1;
    return true;
  }
  badCount += 1;
  console.log('  실패 ' + what);
  console.log('    기대 ' + show(want));
  console.log('    실제 ' + show(got));
  return false;
}

export function checkTrue(what: string, cond: boolean): boolean {
  return check(what, !!cond, true);
}

export function note(text: string): void {
  console.log('  ' + text);
}

export function golden(name: string): string {
  return fs.readFileSync(path.join(GOLDEN, name), 'utf8');
}

/** 요약 한 줄을 찍고 실패 수를 돌려준다. */
export function done(): number {
  console.log(curName + ': 통과 ' + okCount + ' · 실패 ' + badCount);
  return badCount;
}

File: isorpg/lua/tests/harness.lua
-- 세 언어 공통의 아주 작은 테스트 하네스 (루아판).
--
--   프레임워크를 쓰지 않는 이유는 하나다. 같은 테스트를 파이썬·타입스크립트로도
--   돌려야 하는데, 프레임워크가 다르면 출력이 달라지고 출력이 달라지면
--   덱에 실을 로그도 달라진다. 그래서 '이름 · 기대 · 실제' 만 찍는다.
--
--   파이썬과 다른 점 둘.
--   (1) 파이썬은 튜플·리스트를 == 하나로 비교하지만 루아의 == 는 테이블 주소를 본다.
--       그래서 깊은 비교 eq() 를 직접 만든다.
--   (2) 파이썬은 테스트 파일 하나가 곧 프로세스 하나라 sys.exit 로 끝내면 되지만,
--       루아는 run.lua 가 열두 개를 한 프로세스에서 돌린다. 그래서 done() 은
--       standalone 일 때만 os.exit 하고, 아닐 때는 셈만 넘긴다.

local H = {}

H.ok = 0
H.bad = 0
H.name = '?'
H.total_ok = 0
H.total_bad = 0
H.summaries = {}
H.standalone = true

-- 저장소 뿌리 찾기. 파이썬은 __file__ 을 쓰지만 루아 5.1 에는 그런 것이 없다.
local function golden_dir()
  local root = os.getenv('ISORPG_ROOT')
  if root and root ~= '' then
    if root:sub(-1) ~= '/' then root = root .. '/' end
    return root .. 'golden/'
  end
  local cands = {'../golden/', 'golden/', '../../golden/', '../../../golden/'}
  for i = 1, #cands do
    local f = io.open(cands[i] .. 'palette.txt', 'r')
    if f then f:close(); return cands[i] end
  end
  return '../golden/'
end
H.GOLDEN = golden_dir()

local function eq(a, b)
  if a == b then return true end
  if type(a) ~= 'table' or type(b) ~= 'table' then return false end
  for k, v in pairs(a) do
    if not eq(v, b[k]) then return false end
  end
  for k in pairs(b) do
    if a[k] == nil then return false end
  end
  return true
end
H.eq = eq

local function repr(v, depth)
  depth = depth or 0
  local t = type(v)
  if t == 'string' then
    if #v > 160 then
      return string.format('%q', v:sub(1, 160)) .. '...(' .. #v .. '바이트)'
    end
    return string.format('%q', v)
  end
  if t ~= 'table' then return tostring(v) end
  if depth > 2 then return '{...}' end
  local n = #v
  local parts = {}
  local shown = n < 24 and n or 24
  for i = 1, shown do parts[#parts + 1] = repr(v[i], depth + 1) end
  if n > shown then parts[#parts + 1] = '...(' .. n .. '개)' end
  if n == 0 then
    local keys = {}
    for k in pairs(v) do keys[#keys + 1] = tostring(k) end
    table.sort(keys)
    if #keys == 0 then return '{}' end
    return '{키 ' .. table.concat(keys, ',', 1, math.min(#keys, 8)) .. '}'
  end
  return '{' .. table.concat(parts, ', ') .. '}'
end
H.repr = repr

function H.title(name)
  H.name = name
  H.ok = 0
  H.bad = 0
  print(string.format('== %s ==', name))
end

function H.check(what, got, want)
  if eq(got, want) then
    H.ok = H.ok + 1
    return true
  end
  H.bad = H.bad + 1
  print('  실패 ' .. what)
  print('    기대 ' .. repr(want))
  print('    실제 ' .. repr(got))
  return false
end

function H.check_true(what, cond)
  return H.check(what, cond and true or false, true)
end

function H.note(fmt, ...)
  if select('#', ...) > 0 then
    print('  ' .. string.format(fmt, ...))
  else
    print('  ' .. fmt)
  end
end

function H.golden(name)
  local f = io.open(H.GOLDEN .. name, 'r')
  if not f then error('골든 파일을 열 수 없다: ' .. H.GOLDEN .. name) end
  local s = f:read('*a')
  f:close()
  return s
end

-- 문자열을 줄 배열로. 파이썬의 rstrip('\n').split('\n') 과 같다.
function H.lines(text)
  local out = {}
  for line in (text .. '\n'):gmatch('([^\n]*)\n') do out[#out + 1] = line end
  while #out > 0 and out[#out] == '' do out[#out] = nil end
  return out
end

-- 파이썬 테스트가 쓰는 `rs = (1103515245*rs + 12345) % 2^31` 을 그대로 옮긴 것.
-- 그냥 곱하면 1103515245 * 2^31 ~ 2^61 이라 배정밀도를 넘는다. 상·하위로 쪼갠다.
function H.lcg31(rs)
  local h = math.floor(rs / 65536)
  local l = rs - h * 65536
  local hh = 1103515245 * h
  hh = hh - 32768 * math.floor(hh / 32768)          -- mod 2^15 (2^15 * 2^16 = 2^31)
  local v = hh * 65536 + 1103515245 * l + 12345
  return v - 2147483648 * math.floor(v / 2147483648)
end

function H.done()
  print(string.format('%s: 통과 %d · 실패 %d', H.name, H.ok, H.bad))
  H.total_ok = H.total_ok + H.ok
  H.total_bad = H.total_bad + H.bad
  H.summaries[#H.summaries + 1] = {H.name, H.ok, H.bad}
  if H.standalone then
    os.exit(H.bad > 0 and 1 or 0)
  end
end

return H
